editor/plugins/gradient_editor_plugin.h

```diff
@@ -33,26 +33,102 @@
 
 #include "editor/editor_inspector.h"
 #include "editor/editor_plugin.h"
-#include "gradient_editor.h"
 
-class GradientReverseButton : public BaseButton {
-	GDCLASS(GradientReverseButton, BaseButton);
+class EditorSpinSlider;
+class ColorPicker;
+class PopupPanel;
+class GradientTexture1D;
 
-	int margin = 2;
+class GradientEdit : public Control {
+	GDCLASS(GradientEdit, Control);
 
+	Ref<Gradient> gradient;
+	Ref<GradientTexture1D> preview_texture;
+
+	PopupPanel *popup = nullptr;
+	ColorPicker *picker = nullptr;
+
+	bool snap_enabled = false;
+	int snap_count = 10;
+
+	enum GrabMode {
+		GRAB_NONE,
+		GRAB_ADD,
+		GRAB_MOVE
+	};
+
+	GrabMode grabbing = GRAB_NONE;
+	float pre_grab_offset = 0.5;
+	int pre_grab_index = -1;
+	int selected_index = -1;
+	int hovered_index = -1;
+
+	// Make sure to use the scaled values below.
+	const int BASE_SPACING = 4;
+	const int BASE_HANDLE_WIDTH = 8;
+
+	int draw_spacing = BASE_SPACING;
+	int handle_width = BASE_HANDLE_WIDTH;
+
+	int _get_gradient_rect_width();
+
+	void _color_changed(const Color &p_color);
+	void _redraw();
+
+	int _get_point_at(int p_xpos);
+	int _predict_insertion_index(float p_offset);
+	void show_color_picker();
+
+protected:
+	virtual void gui_input(const Ref<InputEvent> &p_event) override;
+	void _notification(int p_what);
+	static void _bind_methods();
+
+public:
+	void set_gradient(const Ref<Gradient> &p_gradient);
+	const Ref<Gradient> get_gradient();
+
+	const ColorPicker *get_picker();
+	const PopupPanel *get_popup();
```

Comment:
`const` goes at the end of the method signature to indicate that it's no mutating.

Reply:
If both are wanted, this can possibly be `const PopupPanel *get_popup() const;`.

BTW the `get_gradient()` one should likely be `const Ref<Gradient> &`.

Reply:
Oh sorry, I honestly wasn't sure what you meant (and what I was doing :sweat:).

Reply:
Can I get the exact syntax I need for the Ref<Gradient> line and for the one in the cpp if it needs changing, plz?

Reply:
It can be `const Ref<Gradient> &get_gradient() const;`